### PR TITLE
Name resolution: keep type vars in subsequent checks

### PR DESCRIPTION
**Pull request by auduchinok**
This is another approach for reporting type arguments, previously experimented in #14773 and #15628. In the following cases I'd like to make it possible to see the checked/inferred types of `I` type arguments:

```fsharp
I<int>
I<_>
```

The previous approaches tried to overwrite reported symbols after the type arguments were checked, but it was quite ad-hoc and didn't work well with type parameters flow in cases like this one:

```fsharp
type I<'T> =
    abstract M: 'T -> unit

{ new I<_> with
      member this.M(i: int) = () }
```

With that approach we'd need this:
* re-report types after type arguments are checked
* find a performant way to replace previously reported type symbol that lacked inferred type arguments

This PR uses a different approach. It turns out that type variables were already always created for symbol reporting, but they were discarded afterwards and new type variables were created for subsequent type application checking. This is the reason why the initially reported type variables lacked inferred type solution: they never participated in the subsequent checks.

This PR makes the initially reported type variables flow through the subsequent type checking phases, so they have the correct solutions after the checking. This works for reporting types first and checking arguments later, because storing the type inference solutions uses mutation. With this approach, the type instantiations are correctly preserved in the cases like the one above.

This PR is not ready yet, and I want to check what tests fail in CI.

### CODE REVIEW

**Comment by auduchinok** [comment]
This is ready. Type arguments are preserved properly now in the captured items. 🎉 

However, they aren't accessible via FCS APIs in some cases, so one of the added tests is disabled for now. I'm going to try to rework the type arguments story for `FSharpEntity` symbols, so 1) type arguments are accessible and 2) less data is duplicated, but it's going to be a separate PR or several of them.

**Review by T-Gro — APPROVED** [review]

File: src/Compiler/Checking/CheckExpressions.fs [hunk 1]
```diff
@@ -4954,16 +4954,22 @@ and TcTypeApp (cenv: cenv) newOk checkConstraints occ env tpenv m tcref pathType
...
-    // If we're not checking constraints, i.e. when we first assert the super/interfaces of a type definition, then just
-    // clear the constraint lists of the freshly generated type variables. A little ugly but fairly localized.
-    if checkConstraints = NoCheckCxs then tps |> List.iter (fun tp -> tp.SetConstraints [])
     let synArgTysLength = synArgTys.Length
     let pathTypeArgsLength = pathTypeArgs.Length
     if tinst.Length <> pathTypeArgsLength + synArgTysLength then
         error (TyconBadArgs(env.DisplayEnv, tcref, pathTypeArgsLength + synArgTysLength, m))
 
+    let tps = tinst |> List.skip pathTypeArgsLength |> List.map (fun t ->
+        match t with
```
Comment (vzarytovskii) [inline src/Compiler/Checking/CheckExpressions.fs:4963]: We shouldn't probably match on ttype directly, need do stripping first

Comment (T-Gro) [reply src/Compiler/Checking/CheckExpressions.fs:4963]: I am not sure of this holds for a type instantiation, but it is a good point to add a test case with a type alias as a typar to see what it does.

If we stripped, we would lose the type alias and then it couldn't be searched for as a separate symbol... ?

Comment (auduchinok) [reply src/Compiler/Checking/CheckExpressions.fs:4963]: I'm not sure it'd be what we want to achieve here. I'm unwrapping the type variables produced by name resolution logic here, I probably don't want to change them in any way. Another approach could be to refactor more places and to pass the typar instantiation explicitly, but it seems that simply unwrapping it like this works for the cases we're interested in here, i.e. to make the types flow into subsequent checking phases after reporting.

File: docs/release-notes/.FSharp.Compiler.Service/8.0.200.md [hunk 2]
```diff
@@ -14,8 +14,9 @@
 * Parser recovers on unfinished enum case declarations. ([PR #16401](https://github.com/dotnet/fsharp/pull/16401))
 * Parser recovers on unfinished record declarations. ([PR #16357](https://github.com/dotnet/fsharp/pull/16357))
 * `MutableKeyword` to [SynFieldTrivia](../reference/fsharp-compiler-syntaxtrivia-synfieldtrivia.html) ([PR #16357](https://github.com/dotnet/fsharp/pull/16357))
+* Name resolution: keep type vars in subsequent checks ([PR #16456](https://github.com/dotnet/fsharp/pull/16456))
```
[the remark this answers is not in the document]
Comment (auduchinok) [reply docs/release-notes/.FSharp.Compiler.Service/8.0.200.md]: @vzarytovskii Thanks, moved.

**Comment by auduchinok** [comment]
Could this get merged, please? 🙂

**Pull request merged**